### PR TITLE
snapstate/check_snap: add snap_microk8s to shared system-usernames

### PR DESCRIPTION
**Pull request by anonymouse64**
This adds support for snaps (specifically microk8s right now) to declare that they want to use a snap_microk8s user/group.

Marking as blocked because we likely want review-tools to not allow any snap other than microk8s (and maybe the juju snap?) to use this group. 

This is basically https://github.com/snapcore/snapd/pull/9396, but s/docker/microk8s/ since our needs for microk8s are greater right now than for docker. We can always add snap_docker afterwards with the next gid.

### CODE REVIEW

File: overlord/snapstate/check_snap.go [hunk 1]
```diff
@@ -176,7 +176,8 @@ var featureSet = map[string]bool{
 // https://docs.docker.com/engine/security/userns-remap/
 // https://github.com/lxc/lxd/blob/master/doc/userns-idmap.md
 var supportedSystemUsernames = map[string]uint32{
-	"snap_daemon": 584788,
+	"snap_daemon":   584788,
+	"snap_microk8s": 584789,
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go:180]: if the snap is not installed with dangerous we really need to check for the snap-id to allow for the use of this user, we cannot just have the store-side check.

Comment (anonymouse64) [reply overlord/snapstate/check_snap.go:180]: Why is the store side check not enough? If I try to upload my foo snap that uses snap_microk8s system-username, it gets rejected the same way that uploading a snap that tries to plug k8s-support interface, etc

Basically, why is snap_microk8s more dangerous that k8s-support ?

Comment (pedronis) [reply overlord/snapstate/check_snap.go:180]: why do you say more?  is about the same, k8s-support is superprivileged so you need either dangerous or a snap-declaration. We could make this dependent on an interface instead of snapd-id but that adds quite  a bit complexity, because the code here is too early to check for what is connected.

Comment (anonymouse64) [reply overlord/snapstate/check_snap.go:180]: We don't have the same snap-id restriction on k8s-support that you are proposing for snap_microk8s, unless you mean to have the equivalent of `allow-installation: false` for system-usernames as well

Comment (pedronis) [reply overlord/snapstate/check_snap.go:180]: that's right, but notwithstanding what I might be hinted at, we have a namespace management issue here, and whatever the final mechanism via the store I don't expect we would give obviously snap_microk8s to everything with k8s-support, unless we think what we want is really snap_k8s ? or snap_kubernetes group? even then as I said because of namespace mgmt I would expect fewer snaps (only the one?) to own the group vs the interface

Comment (anonymouse64) [reply overlord/snapstate/check_snap.go:180]: That is a fair question about whether we should actually make it `snap_k8s`, there is some overlap between the microk8s snap and other snaps (like the juju snap I think?), perhaps @mardy can inquire about that.

Comment (joedborg) [reply overlord/snapstate/check_snap.go]: And, for example, the `kubernetes_worker` snap (https://github.com/charmed-kubernetes/snap-kubernetes-worker).

Comment (anonymouse64) [reply overlord/snapstate/check_snap.go]: @joedborg does that snap need to use the same group as setup by juju or microk8s though? I was under the impression that snap was meant as an isolated worker node that you run somewhere and it registers itself with the control plane that is on a different machine

File: overlord/snapstate/check_snap.go [hunk 2]
```diff
@@ -613,10 +622,20 @@ var osutilEnsureUserGroup = osutil.EnsureUserGroup
...
-		if _, ok := supportedSystemUsernames[user.Name]; !ok {
+		systemUserName, ok := supportedSystemUsernames[user.Name]
+		if !ok {
 			return fmt.Errorf(`snap %q requires unsupported system username "%s"`, si.InstanceName(), user.Name)
 		}
 
+		if systemUserName.AllowedSnapIds != nil {
+			// Only certain snaps can use this user; let's check whether ours
+			// is one of these
+			if !strutil.ListContains(systemUserName.AllowedSnapIds, si.SnapName()) {
```
Comment (anonymouse64) [inline overlord/snapstate/check_snap.go:633]: this isn't the check we want, we want to check not the name of the snap, but the snap ID, i.e. 

```
$ snap info microk8s | grep snap-id
snap-id: EaXqgt1lyCaxKaQCU349mlodBkDCXRcg
```

so we need to be checking for `EaXqgt1lyCaxKaQCU349mlodBkDCXRcg` not `microk8s`.

In general though, I'm not really sure myself where we should:

1) store this information about what snap-ids are allowed to be used with what snap-ids
2) where to perform this check

I think that somewhere in check_snap.go is the right place to perform the check if we do end up storing the information here, since that's where we check other stuff about system-usernames and validateSystemUserNames, but if this information about which snap ids can use which system-usernames gets stored elsewhere I think this won't be applicable. 

I defer to @pedronis for an opinion on where to store this information

Comment (mardy) [reply overlord/snapstate/check_snap.go:633]: > this isn't the check we want, we want to check not the name of the snap, but the snap ID, i.e.
> 
> ```
> $ snap info microk8s | grep snap-id
> snap-id: EaXqgt1lyCaxKaQCU349mlodBkDCXRcg
> ```

I don't get such a line when I run the same command, is it because the package I have was not installed from the store?

> so we need to be checking for `EaXqgt1lyCaxKaQCU349mlodBkDCXRcg` not `microk8s`.

But if this ID only comes from the store, maybe we could have a logic where the list of allowed snap is in the form `["name:<snap-name>", ...] ["id:<snap-id>", ...]` and
1. if the snap ID is found in the allowed list of ids (minus the "id:" prefix), then it's allowed, otherwise
2. if the snap has no ID but its name is found in the allowed list of names (minus the "name:" prefix), then it's allowed, or
3. in all other cases: denied.

Otherwise it would mean that sideloaded snaps would not be able to create their users, which would make debugging and testing harder.

> I think that somewhere in check_snap.go is the right place to perform the check if we do end up storing the information here, since that's where we check other stuff about system-usernames and validateSystemUserNames, but if this information about which snap ids can use which system-usernames gets stored elsewhere I think this won't be applicable.

The information will be coming from the snap store, this is just an *ad interim* solution while the store has not been updated to provide this information. Am I misunderstanding you, or are you suggesting for us to wait on a solution implemented in the store?

Comment (anonymouse64) [reply overlord/snapstate/check_snap.go:633]: Personally, I think that having review-tools prevent usage via blocking uploads to the store unless given specific permission is sufficient for this, but I think what @pedronis was going for was more that:

* if a snap is installed locally, it has no snap ID, and as such is allowed to use snap_microk8s by default (this matches how interface auto-connect assertions work, if a snap is installed locally, it is always allowed to install a plug, but no auto-connections happen)
* I suppose if we eventually know that this information will come from the store that is fine to keep it here, I just was more commenting that this file/package specifically doesn't seem like the best place to store policy information since it is kinda obscure and hard to find unless you know to go looking for it

File: overlord/snapstate/check_snap.go (hunk 1 above)
Comment (ktsakalozos) [reply overlord/snapstate/check_snap.go]: There are no plans to have a user group with elevated permissions for any other kubernetes snaps. MicroK8s is special in this regard.

File: overlord/snapstate/check_snap.go [hunk 3]
```diff
@@ -175,8 +176,16 @@ var featureSet = map[string]bool{
...
-var supportedSystemUsernames = map[string]uint32{
-	"snap_daemon": 584788,
+type systemUsername struct {
+	Id uint32
+	// List of snap IDs which are allowed to declare this user or nil if no
+	// such restriction exists
+	AllowedSnapIds []string
+}
+
+var supportedSystemUsernames = map[string]systemUsername{
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go:186]: my suggestion is to move this and the type above for now into  snapstate/policy/system_usernames.go as publicly exported, and import it for use here from there

File: overlord/snapstate/check_snap.go [hunk 4]
```diff
@@ -175,8 +176,16 @@ var featureSet = map[string]bool{
...
+	"snap_daemon":   {Id: 584788},
+	"snap_microk8s": {Id: 584789, AllowedSnapIds: []string{"microk8s"}},
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go:188]: as Ian mentioned this should list instead the snap-id of microk8s

File: overlord/snapstate/check_snap.go [hunk 5]
```diff
@@ -613,10 +622,20 @@ var osutilEnsureUserGroup = osutil.EnsureUserGroup
 
 func validateSystemUsernames(si *snap.Info) error {
 	for _, user := range si.SystemUsernames {
-		if _, ok := supportedSystemUsernames[user.Name]; !ok {
+		systemUserName, ok := supportedSystemUsernames[user.Name]
+		if !ok {
 			return fmt.Errorf(`snap %q requires unsupported system username "%s"`, si.InstanceName(), user.Name)
 		}
 
+		if systemUserName.AllowedSnapIds != nil {
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go:630]: this check then should look use si.SnapID,  and here we should check whether  `&& si.SnapID != ""`

File: overlord/snapstate/check_snap.go (hunk 2 above)
Comment (pedronis) [reply overlord/snapstate/check_snap.go:633]: this would use si.SnapID then

**Review by pedronis — COMMENTED** [review]
my suggestions

File: overlord/snapstate/check_snap.go (hunk 3 above)
Comment (pedronis) [reply overlord/snapstate/check_snap.go:186]: my suggestion breaks because of circular dependencies which I didn't consider, the new suggestion discussed is simply to introduce:  snap/system_usernames.go

File: overlord/snapstate/check_snap_test.go [hunk 6]
```diff
@@ -1030,6 +1031,18 @@ var systemUsernamesTests = []struct {
...
+	sysIDs: "snap_microk8s: shared",
+	scVer:  "dead 2.4.1 deadbeef bpf-actlog",
+	snapID: "some-uninteresting-snap-id",
+	error:  `snap "foo" is not allowed to use the system user "snap_microk8s"`,
+}, {
+	snapID: "EaXqgt1lyCaxKaQCU349mlodBkDCXRcg", // microk8s
+	sysIDs: "snap_microk8s: shared",
+	scVer:  "dead 2.4.1 deadbeef bpf-actlog",
+}, {
+	sysIDs: "snap_microk8s:\n    scope: shared",
```
Comment (pedronis) [inline overlord/snapstate/check_snap_test.go:1044]: maybe a comment here that the snap id is unset so things are allowed

File: snap/system_usernames.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snap
+
+// SupportedSystemUsernames for now contains the hardcoded list of system
```
Comment (pedronis) [inline snap/system_usernames.go:22]: this needs to be attached to the var now

**Review by pedronis — APPROVED** [review]
I haven't looked at the spread tests but the code changes though look good, just a couple small comments. I have unblocked this.

File: snap/system_usernames.go (hunk 7 above)
Comment (mardy) [reply snap/system_usernames.go:29]: Done!

File: overlord/snapstate/check_snap_test.go (hunk 6 above)
Comment (mardy) [reply overlord/snapstate/check_snap_test.go:1045]: Done

File: tests/main/system-usernames-microk8s/task.yaml [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    fi
+
+    echo "Ensure jq is installed"
+    if ! command -v jq; then
+        snap install --devmode jq
+    fi
+
+    echo "Ensure yaml2json is installed"
+    if ! command -v yaml2json; then
+        snap install --devmode remarshal
```
Comment (anonymouse64) [inline tests/main/system-usernames-microk8s/task.yaml:19]: :smile:

File: tests/main/system-usernames-microk8s/task.yaml [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    # Make sure neither snap_microk8s user nor group are created
+    not getent passwd snap_microk8s
+    not getent group snap_microk8s
+
+    echo "Now install the microk8s snap"
+    snap install "microk8s" 2>&1 | MATCH 'microk8s 1.0.* installed'
+
+    # Make sure both snap_microk8s user and group are created
+    getent passwd snap_microk8s
```
Comment (anonymouse64) [inline tests/main/system-usernames-microk8s/task.yaml:69]: should this test delete these users afterward ?

File: tests/lib/store.sh [hunk 10]
```diff
@@ -82,8 +82,8 @@ EOF
 }
 EOF
 
-    fakestore new-snap-declaration --dir "$dir" --snap-decl-json=/tmp/snap-decl.json "$snap_path"
```
Comment (anonymouse64) [inline tests/lib/store.sh:85]: why this change ?

File: tests/main/system-usernames-microk8s/microk8s/meta/snap.yaml [hunk 11]
```diff
@@ -0,0 +1,9 @@
+name: microk8s
```
Comment (anonymouse64) [inline tests/main/system-usernames-microk8s/microk8s/meta/snap.yaml:1]: this snap.yaml and the test-microk8s-username snap both have identical snap.yaml and bin/sh, we can instead use sed to generate a snap.yaml for each one to replace the name here. See where I have done this for example: https://github.com/snapcore/snapd/blob/master/tests/core/netplan/task.yaml

**Review by anonymouse64 — COMMENTED** [review]
a couple nitpicks, but otherwise looks good to me, thanks for the changes here @mardy

File: tests/lib/store.sh (hunk 10 above)
Comment (mardy) [reply tests/lib/store.sh:85]: It's part of https://github.com/snapcore/snapd/pull/10470, which was recently merged. I'll merge the master branch into this branch, and we should see this change go away.

File: tests/main/system-usernames-microk8s/task.yaml (hunk 8 above)
Comment (mardy) [reply tests/main/system-usernames-microk8s/task.yaml:19]: I'm afraid I'm not getting the humour :-)

Is there something wrong or weird in this check?

Comment (anonymouse64) [reply tests/main/system-usernames-microk8s/task.yaml:19]: No I'm just happy to see someone else use remarshal in the tests, I maintain that snap

**Review by anonymouse64 — COMMENTED** [review]
thanks for the most recent changes +1 from me

File: tests/main/system-usernames-microk8s/task.yaml [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    cp "$TESTSLIB"/assertions/developer1.account-key "$STORE_DIR/asserts"
+
+    snap ack "$TESTSLIB/assertions/developer1.account"
+    snap ack "$TESTSLIB/assertions/developer1.account-key"
+
+    # Create a snap which is not entitled to use the microk8s user
+
+    sed "snap/meta/snap.yaml.in" -e "s/SNAPNAME/test-microk8s-username/" > "snap/meta/snap.yaml"
+    snap_path=$("$TESTSTOOLS"/snaps-state pack-local snap)
+    make_snap_installable "$STORE_DIR" "${snap_path}"
```
Comment (anonymouse64) [inline tests/main/system-usernames-microk8s/task.yaml:46]: as per https://github.com/snapcore/snapd/pull/10494, you will need to ack the assertion here manually or otherwise adjust the `make_snap_installable` helper
